File: framework/fit/java/fit-reactor/src/main/java/modelengine/fitframework/flowable/publisher/FilterPublisherDecorator.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2024-2024. All rights reserved.
 */

package modelengine.fitframework.flowable.publisher;

import static modelengine.fitframework.inspection.Validation.notNull;

import modelengine.fitframework.flowable.Publisher;
import modelengine.fitframework.flowable.Subscriber;
import modelengine.fitframework.flowable.Subscription;
import modelengine.fitframework.flowable.operation.AbstractOperation;

import java.util.function.Predicate;

/**
 * 表示 {@link Publisher} 的携带数据过滤的实现。
 *
 * @param <T> 表示发布者中待过滤数据的类型。
 * @author 何天放
 * @since 2024-02-20
 */
public class FilterPublisherDecorator<T> implements Publisher<T> {
    private final Publisher<T> decorated;
    private final Predicate<T> filter;

    public FilterPublisherDecorator(Publisher<T> decorated, Predicate<T> filter) {
        this.decorated = notNull(decorated, "The decorated filter publisher cannot be null.");
        this.filter = notNull(filter, "The filter cannot be null.");
    }

    @Override
    public void subscribe(Subscriber<T> subscriber) {
        this.decorated.subscribe(new FilterOperation<>(this.filter, subscriber));
    }

    private static class FilterOperation<T> extends AbstractOperation<T, T> {
        private final Predicate<T> filter;

        FilterOperation(Predicate<T> filter, Subscriber<T> subscriber) {
            super(subscriber);
            this.filter = filter;
        }

        @Override
        protected void consume0(Subscription subscription, T data) {
            final boolean result;
            try {
                result = this.filter.test(data);
            } catch (Exception cause) {
                this.getPreSubscription().cancel();
                this.getNextSubscriber().fail(cause);
                return;
            }
            if (result) {
                this.getNextSubscriber().consume(data);
            } else {
                this.getPreSubscription().request(1);
            }
        }
    }
}

File: framework/fit/python/fit_flowable/filt_operation/filter_publisher_decorator.py
# -- encoding: utf-8 --
# Copyright (c) 2024 Huawei Technologies Co., Ltd. All Rights Reserved.
# This file is a part of the ModelEngine Project.
# Licensed under the MIT License. See License.txt in the project root for license information.
# ======================================================================================================================
"""
功 能：对于数据进行过滤的发布者的实现。
"""
from typing import Callable, Any

from fit_flowable.core.publisher import Publisher
from fit_flowable.core.subscriber import Subscriber
from fit_flowable.operation.abstract_operation import AbstractOperation
from fit_flowable.pubilsher.abstract_publisher import AbstractPublisher


class FilterPublisherDecorator(AbstractPublisher):
    """
    对于数据进行过滤的发布者的实现。
    """

    class FilterOperation(AbstractOperation):
        def __init__(self, filter_: Callable[[Any], bool], subscriber: Subscriber):
            super().__init__(subscriber)
            self._filter: Callable[[Any], bool] = filter_
            self._subscriber: Subscriber = subscriber

        def do_consume(self, data) -> None:
            try:
                result = self._filter(data)
            except Exception as cause:
                self.get_pre_subscription().cancel()
                self.get_next_subscriber().fail(cause)
                return
            if result:
                self.get_next_subscriber().consume(data)
            else:
                self.get_pre_subscription().request(1)

    def __init__(self, decorated: Publisher, filter_: Callable[[Any], bool]):
        """
        构造将元素进行过滤的发布者。

        :param decorated: 待进行元素转换的发布者。
        :param filter_: 指定的元素转换方式。
        """
        self._decorated = decorated
        self._filter = filter_

    def do_subscribe(self, subscriber: Subscriber) -> None:
        self._decorated.subscribe(FilterPublisherDecorator.FilterOperation(self._filter, subscriber))
